Make integration test able to generate a jUnit report

jhadvig opened the pull request:
By adding starting the integration tests with `JUNIT_REPORT` variable, all tests will be run in and produce `intergration-test.log`, which in the end will be used by `junitreport` tool to generate a jUnit report - `report.xml`  

@stevekuznetsov PTAL

jhadvig commented:
Also I've figured out that the coverage is not necessary for producing the jUnit report, so I we dont need to recompile the `integration.test` binary

stevekuznetsov commented on hack/test-integration.sh:
```diff
@@ -96,16 +101,37 @@ loop="${TIMES:-1}"
-			ret=1
-		fi
-	done
+        for((i=0;i<${loop};i+=1)); do
```
Was this just indentation? Please put in another commit for easier review

jhadvig replied:
fixed... my IDE screwed me over :-/

stevekuznetsov commented on hack/test-integration.sh:
```diff
@@ -96,16 +101,37 @@ loop="${TIMES:-1}"
+test_result_status="${test_result}        github.com/openshift/origin/test/integration    $(($TESTDURATION / 1000)).$(($TESTDURATION % 1000))s"
+if [[ -n "${test_output_file}" ]]; then
+	echo "${test_result_status}" >> "${test_output_file}"
+	cat ${test_output_file} | junitreport --type gotest --suites nested --stream --roots "github.com/openshift/origin/test/integration" --output "${junit_report_file}"
```
I would refactor `os::test::junit::generate_oscmd_report` to call something like `os::test::junit::generate_report` and write a new function `os::test::junit::generate_gotest_report` that you would trap just like we trap the `oscmd` stuff. Would be useful to share that with `hack/test-go.sh` as well.

jhadvig commented:
No we dont need to add the `JUNIT_REPORT` env since we already provide it when running the tests in https://github.com/openshift/aos-cd-jobs/blob/master/sjb/config/test_cases/test_branch_origin_integration.yml#L10

jhadvig commented:
@stevekuznetsov I've updated the PR.
Created the `os::test::junit::generate_report` function which takes the `--type` as an argument and is called by `os::test::junit::generate_oscmd_report`. 
Also refactored the `test-integration.sh` and `test-go.sh` to use the `os::test::junit::generate_report` function. Didnt wanted to create `os::test::junit::generate_gotest_report` function, since we are not using other functions like `os::test::junit::reconcile_output`, `os::test::junit::check_test_counters`, etc. and both  `test-integration.sh` and `test-go.sh` were just calling the junitreport tool

stevekuznetsov commented on hack/test-integration.sh:
```diff
@@ -96,16 +101,39 @@ loop="${TIMES:-1}"
 # hack/test-integration.sh "(WatchBuilds|Template)"
 tests=( $(go run "${OS_ROOT}/hack/listtests.go" -prefix="${OS_GO_PACKAGE}/${package}.Test" "${testexec}" | grep -E "${1-Test}") )
+
+if [[ -n "${junit_report}" ]]; then
```
The convention is to use `$JUNIT_REPORT` to set these like: https://github.com/openshift/origin/blob/master/hack/test-cmd.sh#L79-L82

stevekuznetsov commented on hack/test-integration.sh:
```diff
@@ -96,16 +101,39 @@ loop="${TIMES:-1}"
+
+if [[ -n "${junit_report}" ]]; then
+	tmp_dir="/tmp/openshift/test-integration"
+	mkdir -p ${tmp_dir}
```
Just use pre-existing temp directories:

```sh
export JUNIT_REPORT_OUTPUT="${LOG_DIR}/raw_test_output.log"
```

stevekuznetsov commented on hack/test-integration.sh:
```diff
@@ -96,16 +101,39 @@ loop="${TIMES:-1}"
+	tmp_dir="/tmp/openshift/test-integration"
+	mkdir -p ${tmp_dir}
+	test_output_file="${tmp_dir}/test-go.log"
+	test_error_file="${tmp_dir}/test-go-err.log"
```
We're not using this for anything in here -- if you are trying to replicate the checks we have in `hack/test-go.sh` for panics, etc -- factor those, too, into common functions.

stevekuznetsov replied:
These would be helpful inside of a `os::test::junit::generate_gotest_report`

stevekuznetsov commented on hack/test-integration.sh:
```diff
@@ -96,16 +101,39 @@ loop="${TIMES:-1}"
 ret=0
+test_result="ok"
 pushd "${OS_ROOT}/${package}" &>/dev/null
+os::log::info "Running \`go test\`..."
```
Not sure this a super useful info

stevekuznetsov commented on hack/test-integration.sh:
```diff
@@ -96,16 +101,39 @@ loop="${TIMES:-1}"
+TESTDURATION=$(($TESTENDTIME - $TESTSTARTTIME))
+test_result_status="${test_result}        github.com/openshift/origin/test/integration    $(($TESTDURATION / 1000)).$(($TESTDURATION % 1000))s"
+if [[ -n "${test_output_file}" ]]; then
+	echo "${test_result_status}" | tee "${test_output_file}"
```
Pipe to `tee` is weird -- use `>>`

stevekuznetsov commented on hack/test-integration.sh:
```diff
@@ -96,16 +101,39 @@ loop="${TIMES:-1}"
 	done
 done
+TESTENDTIME=$(date +%s%3N)
+TESTDURATION=$(($TESTENDTIME - $TESTSTARTTIME))
```
Convention is use to use `$lower_case_variables` for those you're not exporting

stevekuznetsov commented on hack/test-integration.sh:
```diff
@@ -96,16 +101,39 @@ loop="${TIMES:-1}"
+	echo "${test_result_status}" | tee "${test_output_file}"
+	JUNIT_REPORT_OUTPUT="${test_output_file}" ARTIFACT_DIR="${tmp_dir}" os::test::junit::generate_report "gotest"
+else
+	echo "${test_result_status}"
```
Not sure this is a super useful message either -- if you set `$JUNIT_REPORT_OUTPUT` in the `if [[ -n "${JUNIT_REPORT:-}" ]]` stanza like we do for other tests, and use `$ARTIFACT_DIR` from the setup scripts, this could look like:

```sh
echo "${test_result}        github.com/openshift/origin/test/integration    $((test_duration / 1000)).$((test_duration % 1000))s" >> "${JUNIT_REPORT_OUTPUT:-/dev/null}"
os::test::junit::generate_report "gotest"
```

jhadvig commented:
@stevekuznetsov comments addressed. Thanks !

stevekuznetsov commented on hack/test-integration.sh:
```diff
@@ -96,16 +101,33 @@ loop="${TIMES:-1}"
 # hack/test-integration.sh "(WatchBuilds|Template)"
 tests=( $(go run "${OS_ROOT}/hack/listtests.go" -prefix="${OS_GO_PACKAGE}/${package}.Test" "${testexec}" | grep -E "${1-Test}") )
+
+if [[ -n "${JUNIT_REPORT:-}" ]]; then
```
Move this up to line 32 for visual clarity -- should execute fine but is weird that it's after the `exectest()` declaration.

stevekuznetsov commented on hack/test-integration.sh:
```diff
@@ -96,16 +101,33 @@ loop="${TIMES:-1}"
 	done
 done
+test_end_time=$(date +%s%3N)
+test_duration=$(($test_end_time - $test_start_time))
```
Don't need and should not use `$` for var expansion inside of `$(( ))`

stevekuznetsov commented:
[test]

stevekuznetsov commented on hack/lib/test/junit.sh:
```diff
@@ -172,10 +172,22 @@ function os::test::junit::generate_oscmd_report() {
     # use the junitreport tool to generate us a report
     os::util::ensure::built_binary_exists 'junitreport'
```
@jhadvig this should be in `os::test::junit::generate_report()`

stevekuznetsov replied:
Same with the `if` statement:

```sh
    if [[ -z "${JUNIT_REPORT_OUTPUT:-}" ||
          -n "${JUNIT_REPORT_OUTPUT:-}" && ! -s "${JUNIT_REPORT_OUTPUT:-}" ]]; then
          # we can't generate a report
          return
    fi
```

jhadvig commented:
@stevekuznetsov thanks for comments and nits ! Should be fixed now

stevekuznetsov commented:
Integation jUnit looks good:

https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_integration/1773/testReport/github/com_openshift_origin_test_integration /

[merge]
